### PR TITLE
a proof that our set of laws/equations is actually complete

### PR DESCRIPTION
**Pull request by nomeata**
https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/A.20final.20end-to-end.20theorem.20in.20Lean

### CODE REVIEW

**Comment by amirlb** [comment]
From testing on my version of the code, looks like the comparison logic could be simpler: no need to compare expression size (can remove `.forks`), and no need for lots of conditions when comparing laws, this works for me:
```
def Law.MagmaLaw.comp (l1 l2 : Law.NatMagmaLaw) : Ordering :=
  (FreeMagma.comp l1.lhs l2.lhs).then (FreeMagma.comp l1.rhs l2.rhs)

def Law.MagmaLaw.is_canonical (law : Law.MagmaLaw Nat) : Bool :=
  law.symm.canonicalize.comp law ≠ .lt
```

**Comment by amirlb** [comment]
Thinking about it some more, the fact the simple code works for comparison is probably because I generate only the canonical variable ordering in the first place. So fewer things need to be checked for filtering after that.

**Comment by nomeata** [comment]
awaiting-review

**Comment by pitmonticone** [comment]
Build failed with throwing the following error: 

```console
error: no such file or directory (error code: 2)
  file: ././././equational_theories/RArray.lean
Some required builds logged failures:
- Computing build jobs
error: build failed
```

**Comment by nomeata** [comment]
classic, sorry

**Comment by pitmonticone** [comment]
Thanks! I'll review in a few moments.

**Review by pitmonticone — APPROVED** [review]
Just quickly golfed a few bits. Thanks a lot! 

I'll merge after CIs

File: equational_theories/Equations/LawsComplete.lean [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        let mid := lb + w'
+        let l' := laws[mid]
+        if l.comp l' = .lt then
+          go lb w' fuel (by omega)
+        else
+          go mid (w-w') fuel (by omega)
+  termination_by structural fuel
+
+/-- The largest used variable. -/
+def FreeMagma.max : FreeMagma Nat → Nat
```
Comment (amirlb) [inline equational_theories/Equations/LawsComplete.lean:86]: I think all these '.max's aren't used

**Pull request merged**